# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [2]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [3]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [4]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [6]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=True, normalize=False)

SalesTraining    N    Y  All
Gender                      
F              212  277  489
M              201  309  510
All            413  586  999

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [7]:
poke = pd.read_csv("pokemon.csv")

In [8]:
poke["Type 1"]. value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [10]:
poke["Type 1"]. value_counts(normalize = True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [12]:
poke["Type 1" and "Type 2"]. value_counts()

Poison      19
Flying      19
Ground       6
Psychic      6
Water        4
Fairy        3
Ice          3
Grass        2
Steel        2
Rock         2
Fighting     1
Name: Type 2, dtype: int64

In [13]:
pd.crosstab(poke["Stage"], poke["Stage"], margins = True, normalize = False)

Stage   1   2   3  All
Stage                 
1      79   0   0   79
2       0  56   0   56
3       0   0  16   16
All    79  56  16  151

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [14]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [15]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [16]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [18]:
poke["Attack"].mean()

72.54966887417218

In [19]:
poke["Defense"].median()

# looking at the entire distribution of defense values, half are above 65 and half are below

65.0

In [28]:
poke["Defense"].min()

5

In [29]:
poke["Defense"].max()

180

In [30]:
poke["Speed"].mode()

# the two most commonly occurring speeds are 70 and 90

0    70
1    90
dtype: int64

In [27]:
poke["Speed"].min()

15

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [31]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [32]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [41]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [43]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [44]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [45]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [46]:
range = poke["HP"].max() - poke["HP"].min()
print(range)

240


In [48]:
poke["SpAtk"].describe()

count    151.000000
mean      67.139073
std       28.534199
min       15.000000
25%       45.000000
50%       65.000000
75%       87.500000
max      154.000000
Name: SpAtk, dtype: float64

In [51]:
np.percentile(poke["Attack"], 80)

# 80% of attack values are less than 95

95.0

In [52]:
np.percentile(poke["Attack"], 15)

# 15% of attack values are less than 45

45.0

In [53]:
poke["SpDef"].median()

65.0

In [54]:
np.percentile(poke["SpDef"], 50)

65.0

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [56]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [57]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [64]:
stats.zscore(dfz["Salary"])

0    -0.269844
1     3.343557
2    -0.247036
3    -0.270553
4    -0.248449
5    -0.299336
6    -0.272476
7    -0.271102
8    -0.244841
9    -0.281547
10   -0.262206
11   -0.298918
12   -0.130470
13   -0.274427
14   -0.278720
15   -0.296672
16   -0.258333
17   -0.281792
18   -0.274047
19   -0.299274
20   -0.273850
21   -0.254051
22   -0.248976
23   -0.275031
24   -0.257347
25   -0.252134
26   -0.279920
27   -0.268989
28    4.100689
29   -0.273906
Name: Salary, dtype: float64

In [65]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [66]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [67]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [68]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [69]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [70]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [71]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [72]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [73]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [74]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [75]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [77]:
baby = pd.read_excel("babies.xlsx")

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [78]:
baby.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

In [80]:
baby.info()

#all but two columns have missing info
#all columns are numeric, but parity and smoke are coded qualitative

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [85]:
baby.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [89]:
baby.dropna(inplace = True)

In [90]:
baby

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0    77.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0    67.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [ ]:
# parity and smoke are coded with 1s and 0s to mean "yes" or "no"

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [103]:
def previous(parity):
    if parity == 1:
        return "Yes"
    if parity == 0:
        return "No"
    else:
        return "Unknown"

# baby["Previously Pregnant"] = np.where("parity" == 1, "Yes", "No")  <-- WHY NO WORK??
# because it should be: 
# baby["Previously Pregnant"] = np.where(baby["parity"] == 1.0, "Yes", "No")

baby["Previously Pregnant"] = baby["parity"].apply(previous)

baby

bwt  gestation  parity   age  height  weight  smoke Previously Pregnant  \
0     120      284.0       0  27.0    62.0   100.0    0.0                  No   
1     113      282.0       0  33.0    64.0   135.0    0.0                  No   
2     128      279.0       0  28.0    64.0    77.0    1.0                  No   
4     108      282.0       0  23.0    67.0   125.0    1.0                  No   
5     136      286.0       0  25.0    62.0    93.0    0.0                  No   
...   ...        ...     ...   ...     ...     ...    ...                 ...   
1231  113      275.0       1  27.0    60.0   100.0    0.0                 Yes   
1232  128      265.0       0  24.0    67.0   120.0    0.0                  No   
1233  130      291.0       0  30.0    65.0    67.0    1.0                  No   
1234  125      281.0       1  21.0    65.0   110.0    0.0                 Yes   
1235  117      297.0       0  38.0    65.0   129.0    0.0                  No   

     Smoker  
0        No  
1        No  
2       Yes  
4       Yes  
5        No  
...     ...  
1231     No  
1232     No  
1233    Yes  
1234     No  
1235     No  

[1174 rows x 9 columns]

In [104]:
baby.loc[baby["Previously Pregnant"]== "Yes"]

# it worked!

bwt  gestation  parity   age  height  weight  smoke Previously Pregnant  \
271   133      276.0       1  22.0    63.0   119.0    0.0                 Yes   
273   104      274.0       1  20.0    62.0   115.0    1.0                 Yes   
276   123      284.0       1  20.0    65.0   120.0    1.0                 Yes   
279   141      319.0       1  20.0    67.0   140.0    1.0                 Yes   
281   113      282.0       1  36.0    59.0   140.0    0.0                 Yes   
...   ...        ...     ...   ...     ...     ...    ...                 ...   
1223  115      262.0       1  23.0    64.0   136.0    1.0                 Yes   
1225  113      287.0       1  29.0    70.0   145.0    1.0                 Yes   
1226  109      244.0       1  21.0    63.0   102.0    1.0                 Yes   
1231  113      275.0       1  27.0    60.0   100.0    0.0                 Yes   
1234  125      281.0       1  21.0    65.0   110.0    0.0                 Yes   

     Smoker  
271      No  
273     Yes  
276     Yes  
279     Yes  
281      No  
...     ...  
1223    Yes  
1225    Yes  
1226    Yes  
1231     No  
1234     No  

[308 rows x 9 columns]

In [98]:
def smoker(smoke):
    if smoke == 1.0:
        return "Yes"
    if smoke == 0.0:
        return "No"
    else:
        return "Unknown"
    
# baby["Smoker"] = np.where("smoke" == 1.0, "Yes", "No") <-- WHY NO WORK??
# because it should be: 
# baby["Smoker"] = np.where(baby["smoke"] == 1.0, "Yes", "No")

baby["Smoker"] = baby["smoke"].apply(smoker)
baby

bwt  gestation  parity   age  height  weight  smoke Previously Pregnant  \
0     120      284.0       0  27.0    62.0   100.0    0.0                  No   
1     113      282.0       0  33.0    64.0   135.0    0.0                  No   
2     128      279.0       0  28.0    64.0    77.0    1.0                  No   
4     108      282.0       0  23.0    67.0   125.0    1.0                  No   
5     136      286.0       0  25.0    62.0    93.0    0.0                  No   
...   ...        ...     ...   ...     ...     ...    ...                 ...   
1231  113      275.0       1  27.0    60.0   100.0    0.0                  No   
1232  128      265.0       0  24.0    67.0   120.0    0.0                  No   
1233  130      291.0       0  30.0    65.0    67.0    1.0                  No   
1234  125      281.0       1  21.0    65.0   110.0    0.0                  No   
1235  117      297.0       0  38.0    65.0   129.0    0.0                  No   

     Smoker  
0        No  
1        No  
2       Yes  
4       Yes  
5        No  
...     ...  
1231     No  
1232     No  
1233    Yes  
1234     No  
1235     No  

[1174 rows x 9 columns]

In [99]:
baby.loc[baby["Smoker"]== "Yes"]

# it worked!

bwt  gestation  parity   age  height  weight  smoke Previously Pregnant  \
2     128      279.0       0  28.0    64.0    77.0    1.0                  No   
4     108      282.0       0  23.0    67.0   125.0    1.0                  No   
9     143      299.0       0  30.0    66.0   136.0    1.0                  No   
11    144      282.0       0  32.0    64.0   124.0    1.0                  No   
12    141      279.0       0  23.0    63.0   128.0    1.0                  No   
...   ...        ...     ...   ...     ...     ...    ...                 ...   
1224  143      281.0       0  28.0    65.0   135.0    1.0                  No   
1225  113      287.0       1  29.0    70.0   145.0    1.0                  No   
1226  109      244.0       1  21.0    63.0   102.0    1.0                  No   
1227  103      278.0       0  30.0    60.0    87.0    1.0                  No   
1233  130      291.0       0  30.0    65.0    67.0    1.0                  No   

     Smoker  
2       Yes  
4       Yes  
9       Yes  
11      Yes  
12      Yes  
...     ...  
1224    Yes  
1225    Yes  
1226    Yes  
1227    Yes  
1233    Yes  

[459 rows x 9 columns]

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [115]:
# make copy of dataset

babyIqr = baby.copy()

In [116]:
# determine IQR and outlier boundaries of birthweight ("bwt") column

q1 = babyIqr["bwt"].quantile(.25)
q3 = babyIqr["bwt"].quantile(.75)

iqr = q3 - q1

outlierBwtLow = q1 - (1.5*iqr)
outlierBwtHigh = q3 + (1.5*iqr)

print("The lower outlier limit for birthweight is: ", outlierBwtLow)
print("The upper outlier limit for birthweight is: ", outlierBwtHigh)

The lower outlier limit for birthweight is:  73.5
The upper outlier limit for birthweight is:  165.5


In [124]:
# locate outliers in birthweight ("bwt") column

bwtOutliers = babyIqr.loc[(babyIqr["bwt"] < outlierBwtLow) | (babyIqr["bwt"] > outlierBwtHigh)].index
print(bwtOutliers)

Int64Index([ 239,  309,  361,  462,  500,  529,  556,  594,  632,  709,  738,
             747,  829,  904,  912,  978, 1021, 1035, 1063, 1065, 1099, 1139,
            1148, 1169],
           dtype='int64')


In [125]:
# drop outliers in birthweight ("bwt") column

babyIqr = babyIqr.drop(bwtOutliers)

In [129]:
babyIqr.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1150 entries, 0 to 1235
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   bwt                  1150 non-null   int64  
 1   gestation            1150 non-null   float64
 2   parity               1150 non-null   int64  
 3   age                  1150 non-null   float64
 4   height               1150 non-null   float64
 5   weight               1150 non-null   float64
 6   smoke                1150 non-null   float64
 7   Previously Pregnant  1150 non-null   object 
 8   Smoker               1150 non-null   object 
dtypes: float64(5), int64(2), object(2)
memory usage: 89.8+ KB


In [130]:
# determine IQR and outliers of gestation column

q1 = babyIqr["gestation"].quantile(.25)
q3 = babyIqr["gestation"].quantile(.75)

iqr = q3 - q1

outlierGstLow = q1 - (1.5*iqr)
outlierGstHigh = q3 + (1.5*iqr)

print("The lower outlier limit for gestation is: ", outlierGstLow)
print("The upper outlier limit for gestation is: ", outlierGstHigh)

gstOutliers = babyIqr.loc[(babyIqr["gestation"] < outlierGstLow) | (babyIqr["gestation"] > outlierGstHigh)].index

print(gstOutliers)

The lower outlier limit for gestation is:  248.0
The upper outlier limit for gestation is:  312.0
Int64Index([   6,    7,   10,   59,   63,   66,  119,  129,  155,  188,  192,
             198,  210,  215,  217,  234,  240,  253,  260,  279,  345,  373,
             394,  440,  460,  484,  511,  523,  630,  685,  699,  710,  726,
             746,  761,  769,  778,  784,  828,  830,  833,  869,  927,  952,
             969, 1002, 1020, 1026, 1074, 1134, 1140, 1142, 1146, 1152, 1172,
            1178, 1199, 1206, 1207, 1217, 1219, 1226],
           dtype='int64')


In [133]:
# drop outliers in gestation column

babyIqr = babyIqr.drop(gstOutliers)

babyIqr.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1088 entries, 0 to 1235
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   bwt                  1088 non-null   int64  
 1   gestation            1088 non-null   float64
 2   parity               1088 non-null   int64  
 3   age                  1088 non-null   float64
 4   height               1088 non-null   float64
 5   weight               1088 non-null   float64
 6   smoke                1088 non-null   float64
 7   Previously Pregnant  1088 non-null   object 
 8   Smoker               1088 non-null   object 
dtypes: float64(5), int64(2), object(2)
memory usage: 85.0+ KB


In [132]:
# determine IQR and outliers of age column

q1 = babyIqr["age"].quantile(.25)
q3 = babyIqr["age"].quantile(.75)

iqr = q3 - q1

outlierAgeLow = q1 - (1.5*iqr)
outlierAgeHigh = q3 + (1.5*iqr)

print("The lower outlier limit for gestation is: ", outlierAgeLow)
print("The upper outlier limit for gestation is: ", outlierAgeHigh)

ageOutliers = babyIqr.loc[(babyIqr["age"] < outlierAgeLow) | (babyIqr["age"] > outlierAgeHigh)].index

print(ageOutliers)

The lower outlier limit for gestation is:  11.0
The upper outlier limit for gestation is:  43.0
Int64Index([633, 1070], dtype='int64')


In [134]:
# drop outliers in age column

babyIqr = babyIqr.drop(ageOutliers)

babyIqr.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1086 entries, 0 to 1235
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   bwt                  1086 non-null   int64  
 1   gestation            1086 non-null   float64
 2   parity               1086 non-null   int64  
 3   age                  1086 non-null   float64
 4   height               1086 non-null   float64
 5   weight               1086 non-null   float64
 6   smoke                1086 non-null   float64
 7   Previously Pregnant  1086 non-null   object 
 8   Smoker               1086 non-null   object 
dtypes: float64(5), int64(2), object(2)
memory usage: 84.8+ KB


In [135]:
# determine IQR and outliers of height column

q1 = babyIqr["height"].quantile(.25)
q3 = babyIqr["height"].quantile(.75)

iqr = q3 - q1

outlierHgtLow = q1 - (1.5*iqr)
outlierHgtHigh = q3 + (1.5*iqr)

print("The lower outlier limit for height is: ", outlierHgtLow)
print("The upper outlier limit for height is: ", outlierHgtHigh)

hgtOutliers = babyIqr.loc[(babyIqr["height"] < outlierHgtLow) | (babyIqr["height"] > outlierHgtHigh)].index

print(hgtOutliers)

The lower outlier limit for height is:  56.0
The upper outlier limit for height is:  72.0
Int64Index([434, 1208], dtype='int64')


In [136]:
# drop outliers in height column

babyIqr = babyIqr.drop(hgtOutliers)

babyIqr.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1084 entries, 0 to 1235
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   bwt                  1084 non-null   int64  
 1   gestation            1084 non-null   float64
 2   parity               1084 non-null   int64  
 3   age                  1084 non-null   float64
 4   height               1084 non-null   float64
 5   weight               1084 non-null   float64
 6   smoke                1084 non-null   float64
 7   Previously Pregnant  1084 non-null   object 
 8   Smoker               1084 non-null   object 
dtypes: float64(5), int64(2), object(2)
memory usage: 84.7+ KB


In [137]:
# determine IQR and outliers of weight column

q1 = babyIqr["weight"].quantile(.25)
q3 = babyIqr["weight"].quantile(.75)

iqr = q3 - q1

outlierWgtLow = q1 - (1.5*iqr)
outlierWgtHigh = q3 + (1.5*iqr)

print("The lower outlier limit for height is: ", outlierWgtLow)
print("The upper outlier limit for height is: ", outlierWgtHigh)

wgtOutliers = babyIqr.loc[(babyIqr["weight"] < outlierWgtLow) | (babyIqr["weight"] > outlierWgtHigh)].index

print(wgtOutliers)

The lower outlier limit for height is:  79.5
The upper outlier limit for height is:  171.5
Int64Index([   2,   14,   21,   23,   27,   41,   88,  117,  125,  147,  149,
             162,  180,  181,  183,  222,  287,  310,  314,  344,  411,  426,
             458,  509,  512,  522,  528,  563,  595,  608,  622,  695,  696,
             723,  733,  849,  858,  859,  865,  888,  911,  924,  935, 1008,
            1154, 1167, 1212, 1233],
           dtype='int64')


In [138]:
# drop outliers in weight column

babyIqr = babyIqr.drop(wgtOutliers)

babyIqr.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1036 entries, 0 to 1235
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   bwt                  1036 non-null   int64  
 1   gestation            1036 non-null   float64
 2   parity               1036 non-null   int64  
 3   age                  1036 non-null   float64
 4   height               1036 non-null   float64
 5   weight               1036 non-null   float64
 6   smoke                1036 non-null   float64
 7   Previously Pregnant  1036 non-null   object 
 8   Smoker               1036 non-null   object 
dtypes: float64(5), int64(2), object(2)
memory usage: 80.9+ KB


7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [139]:
baby = babyIqr.copy()

In [140]:
baby["Previously Pregnant"].value_counts()

No     754
Yes    282
Name: Previously Pregnant, dtype: int64

In [141]:
baby["Previously Pregnant"].value_counts(normalize = True)

No     0.727799
Yes    0.272201
Name: Previously Pregnant, dtype: float64

In [143]:
baby["Smoker"].value_counts()

No     631
Yes    405
Name: Smoker, dtype: int64

In [144]:
baby["Smoker"].value_counts(normalize = True)

No     0.609073
Yes    0.390927
Name: Smoker, dtype: float64

In [145]:
pd.crosstab(baby["Previously Pregnant"], baby["Smoker"], margins = True, normalize = False)

Smoker                No  Yes   All
Previously Pregnant                
No                   459  295   754
Yes                  172  110   282
All                  631  405  1036

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [146]:
baby.mean()

# mean of parity and smoke are meaningless, as they are quantitative

C:\Users\RVHU32~1\AppData\Local\Temp/ipykernel_12284/2144140817.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  baby.mean()


bwt          120.175676
gestation    279.984556
parity         0.272201
age           27.159266
height        64.009653
weight       125.656371
smoke          0.390927
dtype: float64

In [147]:
baby.median()

# median of parity and smoke are meaningless, as they are quantitative

C:\Users\RVHU32~1\AppData\Local\Temp/ipykernel_12284/1881802232.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  baby.median()


bwt          120.0
gestation    280.0
parity         0.0
age           26.0
height        64.0
weight       125.0
smoke          0.0
dtype: float64

In [148]:
baby.mode()

bwt  gestation  parity   age  height  weight  smoke Previously Pregnant  \
0  117      282.0       0  24.0    65.0   130.0    0.0                  No   

  Smoker  
0     No

9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [149]:
baby.describe()

bwt    gestation       parity          age       height  \
count  1036.000000  1036.000000  1036.000000  1036.000000  1036.000000   
mean    120.175676   279.984556     0.272201    27.159266    64.009653   
std      16.404372    11.396404     0.445308     5.723814     2.481215   
min      75.000000   248.000000     0.000000    15.000000    56.000000   
25%     109.750000   273.000000     0.000000    23.000000    62.000000   
50%     120.000000   280.000000     0.000000    26.000000    64.000000   
75%     131.000000   288.000000     1.000000    31.000000    66.000000   
max     165.000000   312.000000     1.000000    43.000000    72.000000   

            weight        smoke  
count  1036.000000  1036.000000  
mean    125.656371     0.390927  
std      16.581046     0.488194  
min      81.000000     0.000000  
25%     113.000000     0.000000  
50%     125.000000     0.000000  
75%     135.000000     1.000000  
max     171.000000     1.000000

In [150]:
range = max - min

baby["Range"] = baby.range()

TypeError: unsupported operand type(s) for -: 'builtin_function_or_method' and 'builtin_function_or_method'